R/glm_marg.R

```diff
@@ -0,0 +1,133 @@
+#' # Get the effect of `gear` on the outcome value, holding values of `mpg`
+#' # constant
+#' marg(mod, var_interest = 'gear', type = 'effects',
+#'      at = list(mpg = c(15, 21)))
+#'
+#' data(margex)
+#' mod <- glm(outcome ~ as.factor(treatment) + distance,
+#'        data = margex, family = 'binomial')
+#' # Get the level of the outcome variable at different values of `treatment`
+#' marg(mod, var_interest = 'treatment', type = 'levels', at = NULL)
+#' # Get the effect of `treatment` on the outcome variable
+#' marg(mod, var_interest = 'treatment', type = 'effects', at = NULL)
+#' # Get the level of the outcome variable at different values of `distance`
+#' marg(mod, var_interest = 'distance', type = 'levels',
+#'           at = NULL, at_var_interest = c(10, 20, 30))
+#'
+#' # Using a custom variance-covariance matrix for clustered standard errors
+#' # (also requires custom degrees of freedom for T statistic with OLS model),
+#' # clustering on the "arm" variable
+#'
+#' data(margex)
+#' data(cvcov)
+#' # ?cvcov
+#' v <- cvcov$ols$clust
+#' d <- cvcov$ols$stata_dof
+#' mod <- glm(outcome ~ treatment + distance,
+#'            data = margex, family = 'binomial')
+#' marg(mod, var_interest = 'treatment', type = 'levels',
+#'           vcov_mat = v, dof = d)
+#'
+#' # Using weights
+#' data(margex)
+#' mm <- glm(y ~ as.factor(treatment) + age, data = margex, family = 'gaussian',
+#'           weights = distance)
+#' z1 <- marg(mod = mm, var_interest = 'treatment', type = 'levels')[[1]]
+#' z2 <- marg(mod = mm, var_interest = 'treatment', type = 'effects')[[1]]
+#'
+marg.glm <- function(mod, var_interest,
+                     data = mod$data[names(mod$prior.weights), ],
+                     weights = mod$prior.weights,
+                     ...){
+
+  if(all(weights == 1)) weights <- NULL
+
+  .marg(mod = mod, var_interest = var_interest,
+        data = data,
+        weights = weights,
+        ...)
+}
+
+pred_se.glm <- function(model,
+                        deriv_func = model$family$mu.eta,
+                        link_func = model$family$linkinv, ...){
+
+  .pred_se(model = model,
+           deriv_func = model$family$mu.eta,
+           link_func = model$family$linkinv, ...)
+
+}
+
+#' @importFrom stats get_all_vars na.omit
+get_data.glm <- function(model, data, weights){
+
+  # Store original number of rows
+  nrow_orig <- nrow(data)
+
+  # Grab only necessary variables
+  data <- get_all_vars(model, data)
+
+  # Add weights
+  if('_weights' %in% all.vars(model$formula))
+    stop("You cannot use the name '_weights' in the model formula. ",
+         "Please rename to another variable.")
+  if(is.null(weights)) weights <- 1
+  data$`_weights` <- weights
+
+  # Keep completes only
+  data <- na.omit(data)
+
+  # Remove any booleans
+  if(all(data$`T` == TRUE))
+    data$`T` <- NULL
+  if(all(data$`F` == FALSE))
+    data$`F` <- NULL
+
+  # Remove weights
+  weights <- data[['_weights']]
+  data[['_weights']] <- NULL
```

Comment:
We're already stopping if `_weights` is a name in `data`. Can't we just keep this on the dataframe, return the dataframe instead of a list, and then change [the code here](https://github.com/anniejw6/modmarg/pull/124/files?diff=split#diff-587e9fb01bf9b4b2766c48b77d61509eR107) to be lines 102-103 from here?

Anticipating your response - it's a minor change and probably not important, just feels cleaner to me; if you don't agree I'm happy with how it is.

Reply:
Wow. You got my response basically entirely right. I can slightly refactor to make this cleaner, if you'd like.